hatchways/team-bamboo
# Feature: Integration of Profile Listings Page

### Closes #37

### What this PR does (required):
---
### Backend
#### New Routes
##### 1. /profile/sitters -> Get's all sitters, currently does not sort by location, startDate and endData. Sorts logged in user out of results as well.
###### query params
- limit -> number representing the number of sitter profiles to fetch: Default 6.
- page -> number representing the current page: Default 1.
###### response structure
```json
{
    "success": {
        "profiles": [
            {
                "id": "61f224a0153b2fee4cc98624",
                "userId": "61f224a0153b2fee4cc98622",
                "isSitter": true,
                "name": "Norma Byers",
                "description": "Dog sitting, cat sitting, pocket pet and bird card",
                "jobTitle": "Loving pet sitter",
                "address": "Toronto, Ontario",
                "telephone": "",
                "birthday": null,
                "photo": "",
                "hourlyRate": 14,
 "uploadedImages": [],
                "__v": 0
            }
        ]
    }
}
```
#### Minor Updates
##### 1. retrieveUser -> adds user to request object, if the user exists. Does not throw an errors on failure.
##### 2. Profile Model -> Added two new fields
- hourlyRate -> number representing the sitter's hourly rate. Has to be a positive number.
- jobTitle -> string representing the sitter's job title.
---
### Front-end
#### Types
##### 1. Profile -> represents a single profile
###### structure
```ts
interface Profile {
  id: string;
  userId: string;
  isSitter: boolean;
  name?: string;
  description?: string;
  jobTitle?: string;
  gender?: 'male' | 'female' | 'other';
  address?: string;
  telephone?: string;
  birthday: string | null;
  photo?: string;
  hourlyRate?: number;
  uploadedImages?: string[];
  averageRating?: number;
}
```
##### 2. SitterProfile -> represents a profile of type sitter, extends Profile user type.
###### structure
```ts
interface SitterProfile extends Profile {
  isSitter: true;
  hourlyRate: number;
}
```
##### 3. GetSitterProfileParams -> represents the allowed parameters the new route provides.
###### structure
```ts
interface GetSitterProfileParams {
  limit?: number;
  page?: number;
  location?: string;
  start?: string;
  end?: string;
}
```
##### 4. GetSitterProfiles -> represents the return response on a successful call to the api.
###### structure
```ts
interface GetSitterProfiles {
  profiles: SitterProfile[];
}
```
---
#### New API Helpers
##### 1. getSitterProfiles -> helper to get all sitter profiles with provided parameters.
###### arguments
- options -> optional object of type `GetSitterProfileParams`
- controller -> optional parameter to Abort Controller to abort call to backend.
---
#### New Hooks
##### 1. useQuery -> retrieves query params from url.
###### returns
- Object of key/value pairs
###### structure
```ts
interface Query {
  [key: string]: string;
}
```
##### 2. useSittersProfiles -> retrieves all profiles listed as a sitter. Only needed as a requirement for the ProfileListings page.
###### arguments
- limit -> number representing the limit of profiles to retrieve. Default is 6.
- location -> string representing the location of the sitter, not implemented due to time constraints.
- start -> number representing a javascript date in milliseconds.
- to -> number representing a javascript date in milliseconds.
- loadOnMount -> boolean value representing whether or not the profiles should be fetched when hook mounts.
- delay -> number in milliseconds representing the delay after each request.
###### returns
- loadMore -> function which increases the page count and fetches more sitter profiles.
- matchProfiles -> a simple helper function to match over the state, accepts an onError function and a onSuccess function.
- isLoading -> boolean value representing whether or not a fetch request is pending.
---
#### Minor Updates
##### 1. LoadingCard -> Variant of ProfileCard, displays whenever retrieving sitter profiles from the backend.
##### 2. ProfileCard -> Truncate description text so that it is a max of 2 lines.
##### 3. SearchBar -> Added ability to receive starting "query" parameters and attached some additional validation.
##### 4. ProfileListings -> Integrated all of the above changes, including loading animations and query parameter retrieval.
---
### Screenshots / Videos (front-end only):
#### 1. Typical User Experience:
![Usual User Experience](https://user-images.githubusercontent.com/64337339/151308228-0f70cae2-5403-43e6-8925-12e77abde259.gif)

#### 2. Emphasized Loading State
![Loading Screen](https://user-images.githubusercontent.com/64337339/151308289-2f2518b2-2358-493c-90e0-5db6472e516f.gif)

#### 3. Date Picker
![Date Picker](https://user-images.githubusercontent.com/64337339/151308355-138a4de9-c260-4ddf-8c9d-007e7c518647.gif)


### Any information needed to test this feature (required):
- Need to create dummy profiles on the backend.
- ensure the dummy profiles all have the 'isSitter' field set to true.
- ensure the dummy profiles all have an hourly rate mentioned

### Any issues with the current functionality (optional):
- If a user refreshes the page, and they had more than the initial 6 loaded profiles, it will start from the beginning with only 6 profiles loaded.
